### DISCUSSION TITLE
How to "fan out" and execute processing in parallel akin to Step Function distributed map runs?

### DISCUSSION THREAD

**Discussion by whom** [dagster-io/dagster#23391] [Q&A]
I've been tasked with exploring different orchestration alternatives for my team. Right now we're using AWS Step Functions for our data pipelines. It works for the most part, but there are a few things I don't like about it. From my research into Dagster, it seems to solve most of the challenges we face when using Step Functions and I'm in the process of building some POCs to demonstrate to my team in the coming week or two.

A very common pattern we have in our Pipelines is the following:

1. Process A will query an internal database table for some data and return N records. It essentially looks for records that haven't been updated in the past 24 hours.
2. Process B will process each of the N records in parallel, up to a specified concurrency limit. It makes API calls to an external source to get updated information about each record.
3. Process C will take the consolidated results and either do further processing or persist it to a data store.

We were able to accomplish this using [Batch Map Runs](https://docs.aws.amazon.com/step-functions/latest/dg/concepts-examine-map-run.html). In an initial POC I was able to replicate Process B by using `ThreadPoolExecutor` but AFAIK that's not true concurrent processing.

I've scoured Google to try to figure out how to do this, but nothing has worked. These two attempts were the most promising, but neither quite worked the way I expected:

1. Process A would be one job and Processes B and C would be another job. The first job would run, get the records, then update a dynamic partition used by Process B. Then I could run the second job with the dynamically filled partition. I don't like this approach because I have to trigger two separate jobs instead of just 1.
2. My latest attempt uses `SensorResult` similar to [this](https://github.com/dagster-io/dagster/blob/master/examples/assets_dynamic_partitions/assets_dynamic_partitions/release_sensor.py) example. This got what I needed done, but from looking at the logs it seems like the `SensorDaemon` keeps firing this sensor even after I'm done processing them.

I'm also wondering, given that one of the things about partitions is setting cursors and making sure you don't process the same key more than once, whether they even fit my use case since I can and want to process the same record over multiple days and runs. All of which leads me to believe that maybe partitions aren't actually what I want?

Sorry for the long-winded post, but I wanted to share with what I found so far. Does Dagster provide a good way for me to do this fan-out approach for Process B? Or should I just stick with threads (or maybe `asyncio`) since Process B is I/O-bound anyway?

**Comment by robmcd** [comment]
Perhaps this? https://github.com/dagster-io/dagster/discussions/15521
